genobuilder.py

```diff
@@ -461,34 +463,32 @@ def _random_sampling_geno(self, callset, mask_file=None, seed=None):
 
         return chroms, pos, slices
 
-    def _resize_from_ts(self, ts, flip=True):
+    def _resize_from_ts(self, ts, rng, flip=True):
         """Returns a genotype matrix with a fixed number of columns,
         as specified in size"""
 
-        m = np.zeros((ts.num_samples, self.fixed_dim), dtype=float)
-        flip = {0: 1, 1: 0, -1: -1}
+        m = np.zeros((ts.num_samples, self.fixed_dim), dtype=int)
+        ac_thresh = self.maf_thresh * ts.num_samples
 
         for variant in ts.variants():
 
-            # Calculate allele frequency
-            af = np.mean(variant.genotypes)
-
             # Filter by MAF
-            if self.maf_thresh is not None:
-                if af < self.maf_thresh or af > 1 - self.maf_thresh:
-                    continue
+            genotypes = variant.genotypes
+            ac1 = np.sum(genotypes)
+            ac0 = len(genotypes) - ac1
+            if min(ac0, ac1) < ac_thresh:
+                continue
 
-            # Polarise the matrix by major allele frequency.
+            # Polarise 0 and 1 in genotype matrix by major allele frequency.
+            # If allele counts are the same, randomly choose a major allele.
             if flip:
-                if af > 0.5 or (af == 0.5 and random.Random() > 0.5):
-                    variant.genotypes = [flip[b] for b in variant.genotypes]
+                if ac1 > ac0 or (ac1 == ac0 and rng.random() > 0.5):
+                    genotypes ^= 1
 
             j = int(variant.site.position * self.fixed_dim / ts.sequence_length)
-            np.add(
-                m[:, j], variant.genotypes, out=m[:, j], where=variant.genotypes != -1
-            )
+            m[:, j] += genotypes
```

Comment:
I wasn't completely sure about this change. Under what circumstances can the genotypes be `-1`?